[Serve][Doc] Add loadbalancer and monitoring section for K8s (#36775)
ray-project/ray: doc/source/serve/production-guide/kubernetes.md

In most use cases, it is recommended to enable Kubernetes autoscaling to fully utilize the resources in your cluster. If you are using GKE, you can utilize the AutoPilot Kubernetes cluster. For instructions, see [Create an Autopilot Cluster]((https://cloud.google.com/kubernetes-engine/docs/how-to/creating-an-autopilot-cluster)). For EKS, you can enable Kubernetes cluster autoscaling by utilizing the Cluster Autoscaler. For detailed information, see [Cluster Autoscaler on AWS](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md). To understand the relationship between Kubernetes autoscaling and Ray autoscaling, see [Ray Autoscaler with Kubernetes Cluster Autoscaler](kuberay-autoscaler-with-ray-autoscaler).
:::

## Load balancer

> You can setup ingress to expose your Serve application to the internet. See [example](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml)

Reviewer: ```suggestion
You can set up an ingress to expose your Serve application to the internet. See [this configuration](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml) for an example.
```

> You can setup ingress to expose your Serve application to the internet. See [example](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml)

Reviewer 2: ```suggestion
Set up ingress to expose your Serve application to the internet. ([Example])(https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml)
```

Reviewer 2: Hope that syntax works.

> Set up ingress to expose your Serve application to the internet. See [this configuration](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml)

Reviewer 3: What if you want an internal service (not exposed to the internet)?

Author: Rephrased. Basically it depends on network config whether your ingress can be connected by external.

:::{note}

> - Ray Serve provisions an HTTP proxy on every node, allowing you to use `/-/routes` as the endpoint for node health checks.

Reviewer: ```suggestion
- Ray Serve runs an HTTP proxy on every node. Each proxy exposes contains a `/healthz` endpoint that you can query to check its health.
```

> - Ray Serve uses port 8000 as the default HTTP proxy traffic port. You can change the port by setting `http_options` in the Serve config. Learn more details [here](serve-multi-application)

Reviewer 2: ```suggestion
- Ray Serve uses port 8000 as the default HTTP proxy traffic port. You can change the port by setting `http_options` in the Serve config. See [Deploy Multiple Applications](serve-multi-application) for more details.
```

> - Ray Serve uses port 8000 as the default HTTP proxy traffic port. You can change the port by setting `http_options` in the Serve config. Learn more details [here](serve-multi-application)

Reviewer: ```suggestion
- Ray Serve uses port 8000 as the default HTTP proxy traffic port. You can change the port by setting `http_options` in the Serve config. Learn more details [here](serve-multi-application).
```

:::

> ## Monitoring

Reviewer 3: Let's also add how to get to the kuberay logs for troubleshooting

> ## Monitoring

Author: added logging with Loki link and trouble shooting notes.

> ## Monitoring

Reviewer 3: thanks for updating! Do we have anything we recommend for metrics too? Btw I noticed this guide as well - https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/logging.html . @scottsun94 which one do you usually recommend?

> ## Monitoring

Reviewer 4: For log aggregation tools, we list a few [here](https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/logging.html): [Fluent Bit](https://docs.fluentbit.io/manual), [Vector](https://vector.dev/), [Fluentd](https://docs.fluentd.org/), [Filebeat](https://www.elastic.co/guide/en/beats/filebeat/7.17/index.html), and [Promtail](https://grafana.com/docs/loki/latest/clients/promtail/). We use Fluent Bit as an example but any of them should work and it depends on the users' preferences.

Serve doc have instructions for how to use Promtail with Loki which is good. It doesn't expand on how to persist the logs using Promtail. Here is my suggestion for how to reorganize the links in this section:
- people need to learn how to export metrics first
- then about dashboard
- lastly log persistence

```
Learn how to export metrics and set up the Grafana [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/)
Learn more about how to configure and manage Dashboard [here](observability-configure-manage-dashboard).
Learn about the Ray Serve Dashboard [here](serve-monitoring).
Learn about the [Ray Serve logs](serve-logging) and how to [persistent logs](https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/logging.html )on Kubernetes clusters.
```

> ## Monitoring

Author: Updated!

> You can monitor your Serve application using the Ray dashboard.

Reviewer 2: ```suggestion
Monitor your Serve application using the Ray Dashboard.
```

> Learn more about how to configure & manage dashboard [here](observability-configure-manage-dashboard)

Reviewer 2: ```suggestion
Learn more about how to configure and manage Dashboard [here](observability-configure-manage-dashboard)
```

> Learn more about how to configure & manage dashboard [here](observability-configure-manage-dashboard)

Reviewer: ```suggestion
Learn more about how to configure and manage the dashboard [here](observability-configure-manage-dashboard).
```

> Learn ray serve dashboard [here](serve-monitoring)

Reviewer 2: ```suggestion
Learn about the Ray Serve Dashboard [here](serve-monitoring)
```

> Learn ray serve dashboard [here](serve-monitoring)

Reviewer: ```suggestion
Learn more about the dashboard's Ray Serve section [here](serve-monitoring).
```

> Learn how to setup Grafana dashboard [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/)

Reviewer: ```suggestion
Learn how to set up the Grafana dashboard [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/).
```

> Learn how to setup Grafana dashboard [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/)

Reviewer 2: ```suggestion
Learn how to set up Grafana for Dashboard [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/)
```

:::

## Load balancer
Set up ingress to expose your Serve application with a load balancer. See [this configuration](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice-alb-ingress.yaml)

:::{note}
- Ray Serve runs HTTP proxy on every node, allowing you to use `/-/routes` as the endpoint for node health checks.
- Ray Serve uses port 8000 as the default HTTP proxy traffic port. You can change the port by setting `http_options` in the Serve config. Learn more details [here](serve-multi-application).
:::

## Monitoring
Monitor your Serve application using the Ray Dashboard.
- Learn more about how to configure and manage Dashboard [here](observability-configure-manage-dashboard).
- Learn about the Ray Serve Dashboard [here](serve-monitoring).

> - Learn how to set up the Grafana for Dashboard [here](https://ray-project.github.io/kuberay/guidance/prometheus-grafana/)

Reviewer 4: It relies on both Prometheus and Grafana. How about this?
```suggestion
- Learn how to set up [Prometheus](prometheus-setup) and [Grafana](grafana) for Dashboard
```

@sihanwang41 Can we send them to this page instead?  https://docs.ray.io/en/latest/cluster/metrics.html#setting-up-your-prometheus-server

Author: nice, I wasn't aware of we have this content in the ray doc already.

> - Learn about the [Ray Serve logs](serve-logging) and how to [persistent logs](kuberay-logging) on Kubernetes clusters.

Reviewer 4: ```suggestion
- Learn about the [Ray Serve logs](serve-logging) and how to [persistent logs](kuberay-logging) on Kubernetes.
```
